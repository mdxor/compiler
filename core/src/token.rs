#[cfg(test)]
use serde::Serialize;

#[derive(Eq, PartialEq, Debug)]
#[cfg_attr(test, derive(Serialize))]
pub struct Token<T> {
  pub span: Span,
  pub value: T,
}

#[derive(Eq, PartialEq, Debug, Clone)]
#[cfg_attr(test, derive(Serialize))]
pub struct Span {
  pub start: usize,
  pub end: usize,
}

#[derive(Eq, PartialEq, Debug)]
#[cfg_attr(test, derive(Serialize))]
pub enum HeadingLevel {
  H1,
  H2,
  H3,
  H4,
  H5,
  H6,
}

impl HeadingLevel {
  pub fn new(level: usize) -> Option<HeadingLevel> {
    match level {
      1 => Some(HeadingLevel::H1),
      2 => Some(HeadingLevel::H2),
      3 => Some(HeadingLevel::H3),
      4 => Some(HeadingLevel::H4),
      5 => Some(HeadingLevel::H5),
      6 => Some(HeadingLevel::H6),
      _ => None,
    }
  }

  pub fn to_str(&self) -> &str {
    match self {
      HeadingLevel::H1 => "h1",
      HeadingLevel::H2 => "h2",
      HeadingLevel::H3 => "h3",
      HeadingLevel::H4 => "h4",
      HeadingLevel::H5 => "h5",
      HeadingLevel::H6 => "h6",
    }
  }
}

#[derive(Eq, PartialEq, Debug)]
#[cfg_attr(test, derive(Serialize))]
pub struct LinkDefinition<'source> {
  pub label: &'source str,
  pub url: &'source str,
  pub title: String,
}

#[derive(Eq, PartialEq, Debug)]
#[cfg_attr(test, derive(Serialize))]
pub enum Align {
  Left,
  Center,
  Right,
}

#[derive(Eq, PartialEq, Debug)]
#[cfg_attr(test, derive(Serialize))]
pub enum BlockToken {
  Paragraph {
    raws: Vec<Span>,
  },
  JSX(JSXElement),
  ATXHeading {
    raws: Vec<Span>,
    level: HeadingLevel,
  },
  SetextHeading {
    raws: Vec<Span>,
    level: HeadingLevel,
  },
  IndentedCode(Vec<Span>),
  BlankLine,
  ThematicBreak,
  BlockQuote {
    blocks: Vec<Token<BlockToken>>,
    level: usize,
  },
  FencedCode {
    meta_span: Span,
    code_spans: Vec<Span>,
  },
  List {
    ch: u8,
    is_tight: bool,
    order_span: Span,
    blocks: Vec<Token<BlockToken>>,
  },
  ListItem {
    indent: usize,
    blocks: Vec<Token<BlockToken>>,
  },
  // LinkDefinition,
  // Table,
  // TableHead,
  // TableCell(&'source str, bool, bool),
  // TableAlignment,
  // TableAlign(Align),
  // TableRow
}

#[derive(Eq, PartialEq, Debug)]
#[cfg_attr(test, derive(Serialize))]
pub enum InlineToken {
  TextSegment,
  MaybeLinkStart,
  MaybeEmphasis {
    ch: u8,
    repeat: usize,
    can_open: bool,
    can_close: bool,
  },
  EmphasisStart,
  EmphasisEnd,
  LinkStart {
    url: Span,
    title: Vec<Span>,
  },
  LinkEnd,
  //
  Emphasis(Vec<Token<InlineToken>>),
  Link {
    url: Span,
    title: Vec<Span>,
    text_children: Vec<Token<InlineToken>>,
  },
  Text(Vec<Span>),
  Code(Vec<Span>),
  CodeSegment,
  SoftBreak,
  HardBreak,
  // is email
  AutoLink(bool),
  JSX(JSXElement),
}

#[derive(Eq, PartialEq, Debug)]
pub struct AST<T> {
  pub span: Span,
  pub children: Vec<T>,
}

#[derive(Eq, PartialEq, Debug)]
#[cfg_attr(test, derive(Serialize))]
pub enum ContainerBlock {
  BlockQuote(usize),
  List(u8),
  ListItem(usize),
}

#[derive(Eq, PartialEq, Debug)]
#[cfg_attr(test, derive(Serialize))]
pub enum JSToken {
  Keyword(Span),
  Punctuator(Span),
  String(Span),
  Template(Span),
  Identifier(Span),
  Number(Span),
  Text(Span),
}

#[derive(Eq, PartialEq, Debug)]
#[cfg_attr(test, derive(Serialize))]
pub enum JSXAttr {
  Spread(Vec<JSXExpressionSegment>),
  KeyLiteralValue {
    key: Span,
    value: Span,
  },
  KeyValue {
    key: Span,
    value: Vec<JSXExpressionSegment>,
  },
  KeyTrueValue {
    key: Span,
  },
}

#[derive(Eq, PartialEq, Debug)]
#[cfg_attr(test, derive(Serialize))]
pub enum JSXExpressionSegment {
  JS(Span),
  Element(JSXElement),
}

#[derive(Eq, PartialEq, Debug)]
#[cfg_attr(test, derive(Serialize))]
pub enum JSX {
  Element(JSXElement),
  Text(Span),
  Expression(Vec<JSXExpressionSegment>),
}

#[derive(Eq, PartialEq, Debug)]
#[cfg_attr(test, derive(Serialize))]
pub struct JSXElement {
  pub tag: String,
  pub attributes: Vec<JSXAttr>,
  pub children: Vec<JSX>,
}
