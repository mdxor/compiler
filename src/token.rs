#[derive(Debug, PartialEq)]
pub enum Token<'a> {
  Heading1,
  Heading2,
  Heading3,
  Heading4,
  Heading5,
  Heading6,
  Text(&'a str),
  InlineCode,
}

pub fn get_token(token: &str) -> Token {
  match &*token {
    "#" => Token::Heading1,
    "##" => Token::Heading2,
    "###" => Token::Heading3,
    "####" => Token::Heading4,
    "#####" => Token::Heading5,
    "######" => Token::Heading6,
    _ => Token::Text(token),
  }
}
