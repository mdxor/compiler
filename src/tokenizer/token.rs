#[cfg(test)]
use serde::Serialize;

#[derive(Debug, PartialEq)]
#[cfg_attr(test, derive(Serialize))]
pub struct TokenResult<'source> {
  pub size: usize,
  pub token: Token<'source>,
}

#[derive(Debug, PartialEq)]
#[cfg_attr(test, derive(Serialize))]
pub enum Token<'source> {
  ATXHeading(ATXHeading<'source>),
  ThematicBreak,
  SetextHeading(SetextHeading<'source>),
  IndentedCode(IndentedCode<'source>),
  FencedCode(FencedCode<'source>),
}

#[derive(Debug, PartialEq)]
#[cfg_attr(test, derive(Serialize))]
pub struct ATXHeading<'source> {
  pub level: usize,
  pub raw_inlines: &'source str,
}

#[derive(Debug, PartialEq)]
#[cfg_attr(test, derive(Serialize))]
pub struct SetextHeading<'source> {
  pub level: usize,
  pub raw: &'source str,
}

#[derive(Debug, PartialEq)]
#[cfg_attr(test, derive(Serialize))]
pub struct IndentedCode<'source> {
  pub codes: Vec<&'source str>,
}

#[derive(Debug, PartialEq)]
#[cfg_attr(test, derive(Serialize))]
pub struct FencedCode<'source> {
  pub code: &'source str,
  pub meta_string: &'source str,
  pub language: &'source str,
}
