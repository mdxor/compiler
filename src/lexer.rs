use crate::jsx;
use crate::rule::Rule;
use crate::token;
use lazy_static::lazy_static;
use regex::Regex;

lazy_static! {
  static ref RULE: Rule = Rule::new();
}

pub struct Lexer<'a> {
  source: &'a str,
}

impl<'a> Lexer<'a> {
  pub fn new(source: &'a str) -> Self {
    Lexer { source }
  }

  fn skip_whitespace(&mut self) -> usize {
    lazy_static! {
      static ref WHITESPACE_REGEX: Regex = Regex::new("^ +").unwrap();
    }
    if let Some(caps) = WHITESPACE_REGEX.captures(self.source) {
      let size = caps.get(0).unwrap().as_str().len();
      self.move_by(size);
      size
    } else {
      0
    }
  }

  fn move_by(&mut self, size: usize) -> &'a str {
    let result = &self.source[..size];
    self.source = &self.source[size..];
    result
  }

  pub fn tokenize(&mut self) -> Result<token::AST<'a>, &'a str> {
    let mut blocks = vec![];
    loop {
      if self.source.is_empty() {
        break;
      } else {
        let token = self.scan_block()?;
        blocks.push(token);
      }
    }
    let ast = token::AST { blocks };
    Ok(ast)
  }

  fn scan_block(&mut self) -> Result<token::Block<'a>, &'a str> {
    if RULE.indented_code.is_match(self.source) {
      self.move_by(4);
      return Ok(self.scan_single_line_code());
    } else {
      // <=3 whitespace
      self.skip_whitespace();
      if let Some(caps) = RULE.empty_atx_heading.captures(self.source) {
        let size = caps.get(1).unwrap().as_str().len();
        self.move_by(caps.get(0).unwrap().as_str().len());
        return Ok(token::Block::Leaf(token::LeafBlock::Heading(
          token::Heading {
            level: size as u8,
            inlines: vec![],
          },
        )));
      }
      if let Some(caps) = RULE.atx_heading.captures(self.source) {
        let size = caps.get(1).unwrap().as_str().len();
        return self.scan_heading(size);
      } else if self.source.starts_with("```") {
        return Ok(self.scan_multiple_line_code());
      }
      return self.scan_paragraph();
    }
  }

  fn scan_paragraph(&mut self) -> Result<token::Block<'a>, &'a str> {
    let inlines = self.scan_inlines()?;
    Ok(token::Block::Leaf(token::LeafBlock::Paragraph(
      token::Paragraph { inlines },
    )))
  }

  fn scan_heading(&mut self, level: usize) -> Result<token::Block<'a>, &'a str> {
    self.move_by(level + 1);
    let mut inlines = self.scan_inlines()?;

    // A closing sequence of # characters is optional:
    // # Title ### => <h1>Title<h1/>
    if !inlines.is_empty() {
      let last_inline = inlines.pop().unwrap();
      let mut closing_size = 0;
      if let token::Inline::Text(text) = last_inline {
        if let Some(caps) = RULE.closing_atx_heading.captures(text) {
          closing_size = caps.get(0).unwrap().as_str().len();
          inlines.push(token::Inline::Text(&text[..text.len() - closing_size]));
        }
      }
      if closing_size == 0 {
        inlines.push(last_inline);
      }
    }
    let heading = token::Heading {
      level: level as u8,
      inlines,
    };
    Ok(token::Block::Leaf(token::LeafBlock::Heading(heading)))
  }

  fn scan_inlines(&mut self) -> Result<Vec<token::Inline<'a>>, &'a str> {
    self.skip_whitespace();
    let mut inlines = vec![];
    loop {
      if self.source.is_empty() || self.source.starts_with("\n") {
        if self.source.starts_with("\n") {
          self.move_by(1);
        }
        return Ok(inlines);
      } else if self.source.starts_with("`") {
      } else {
        inlines.append(&mut self.scan_inline_text());
      }
    }
  }

  fn scan_inline_text(&mut self) -> Vec<token::Inline<'a>> {
    let mut chars = self.source.chars();
    let mut text_size = 0;
    let mut texts = vec![];
    let mut escaped = false;
    loop {
      if let Some(char) = chars.next() {
        if !escaped {
          match char {
            '\n' | '`' => {
              texts.push(token::Inline::Text(self.move_by(text_size)));
              break;
            }
            '\\' => {
              escaped = !escaped;
            }
            _ => {
              text_size += char.len_utf8();
            }
          }
        } else {
          match char {
            '\n' => {
              if text_size > 0 {
                texts.push(token::Inline::Text(self.move_by(text_size + 1)));
              }
              break;
            }
            '\\' => {
              if text_size > 0 {
                texts.push(token::Inline::Text(self.move_by(text_size)));
                text_size = 0;
              }
              texts.push(token::Inline::Text(r"\"));
            }
            '`' | '#' => {
              if text_size > 0 {
                texts.push(token::Inline::Text(self.move_by(text_size)));
                text_size = 0;
              }
              self.move_by(1);
              texts.push(token::Inline::Text(self.move_by(1)));
            }
            _ => {
              text_size += char.len_utf8() + 1;
            }
          }
          escaped = false;
        }
      } else {
        if escaped {
          text_size += 1;
        }
        texts.push(token::Inline::Text(self.move_by(text_size)));
        break;
      }
    }
    texts
  }

  fn scan_single_line_by_end_char(&mut self, end_char: char) -> &'a str {
    let mut size = 0;
    let mut chars = self.source.chars();
    loop {
      if let Some(char) = chars.next() {
        if char == end_char || char == '\n' {
          break;
        } else {
          size += char.len_utf8();
        }
      } else {
        return self.move_by(size);
      }
    }
    self.move_by(size + 1)
  }

  fn scan_single_line_code(&mut self) -> token::Block<'a> {
    let code = self.scan_single_line_by_end_char('\n');
    token::Block::Leaf(token::LeafBlock::IndentedCode(code))
  }

  fn scan_multiple_line_code(&mut self) -> token::Block<'a> {
    lazy_static! {
      static ref CODE_END_REGEX: Regex = Regex::new(r"(^ {0,3}|\n {0,3})``` *\n?").unwrap();
    }
    self.move_by(3);
    let mut language = "";
    let mut metastring = "";
    let mut code = "";

    self.skip_whitespace();
    let mut code_info = self.scan_single_line_by_end_char('\n');
    if code_info.ends_with("\n") {
      code_info = &code_info[..code_info.len() - 1];
    }
    if !code_info.is_empty() {
      if let Some(index) = code_info.find(" ") {
        language = &code_info[0..index].trim();
        metastring = &code_info[index..].trim_start();
      } else {
        language = code_info.trim();
      }
    }

    if let Some(captures) = CODE_END_REGEX.captures(self.source) {
      let end_token = captures.get(0).unwrap().as_str();
      let code_size = self.source.find(end_token).unwrap();
      code = self.move_by(code_size);
      self.move_by(end_token.len());
    } else {
      code = self.move_by(self.source.len());
    }
    token::Block::Leaf(token::LeafBlock::FencedCode(token::FencedCode {
      code,
      language,
      metastring,
    }))
  }

  fn scan_jsx(&mut self, is_inline: bool) -> Result<jsx::JSXNode<'a>, &'a str> {
    let mut jsx_parser = jsx::JSXParser::new(self.source, 0, is_inline);
    let jsx_node = jsx_parser.parse();
    self.move_by(jsx_parser.size);
    jsx_node
  }
}
