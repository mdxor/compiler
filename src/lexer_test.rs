use crate::lexer;
use crate::token;
fn tokenizes(cases: Vec<&str>) -> Vec<Result<token::AST, &str>> {
  let mut results = vec![];
  for case in &cases {
    let mut lex = lexer::Lexer::new(case);
    let ast = lex.tokenize();
    results.push(ast)
  }
  results
}

#[test]
fn test_indented_code() {
  let cases = vec!["    abc", "    <div></div>"];
  insta::assert_yaml_snapshot!(tokenizes(cases));
}
#[test]
fn test_fenced_code() {
  let cases = vec![
    "```\ncode\n```",
    "```jsx\nlet a = 11;\n```",
    "```jsx meta\nlet a = 11;\n```",
  ];
  insta::assert_yaml_snapshot!(tokenizes(cases));
}
#[test]
fn test_atx_heading() {
  let cases = vec![
    // "# foo",
    // "## foo",
    // "### foo",
    // "#### foo",
    // "##### foo",
    // "###### foo",
    // "####### foo",
    // "#5 bolt",
    // "#hashtag",
    // r"\## foo",
    // // TODO:
    // // r"# foo *bar* \*baz\*"
    // "#                  foo                     ",
    // "### foo",
    // " ## foo",
    // "  # foo",
    // "    # foo",
    // // "foo\n    # bar",
    // "## foo ##\n###   bar    ###",
    // "# foo ##################################\n##### foo ##",
    // "### foo ###     ",
    // "### foo ### b",
    // "# foo#",
    // "****\n## foo\n****"
    "Foo bar\n# baz\nBar foo",
    // "## \n#\n### ###",
  ];
  insta::assert_yaml_snapshot!(tokenizes(cases));
}
#[test]
fn test_thematic_break() {
  let cases = vec![
    "***\n---\n___",
    "+++",
    "===",
    "--\n**\n__",
    "***\n ***\n   ***",
    "    ***",
    // "Foo\n    ***",
    "_____________________________________",
    "- - -",
    " **  * ** * ** * **",
    "-     -      -      -",
    "- - - -    ",
    "_ _ _ _ a\na------\n---a---",
    // " *-*",
    // "- foo\n***\n- bar"
    "Foo\n***\nbar",
    // "Foo\n---\nbar",
    // "* Foo\n* * *\n* Bar",
    // "- Foo\n- * * *",
  ];
  insta::assert_yaml_snapshot!(tokenizes(cases));
}
