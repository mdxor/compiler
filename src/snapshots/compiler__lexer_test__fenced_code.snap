---
source: src/lexer_test.rs
expression: tokenizes(cases)

---
- Ok:
    blocks:
      - Leaf:
          FencedCode:
            code: code
            language: ""
            metastring: ""
- Ok:
    blocks:
      - Leaf:
          FencedCode:
            code: let a = 11;
            language: jsx
            metastring: ""
- Ok:
    blocks:
      - Leaf:
          FencedCode:
            code: let a = 11;
            language: jsx
            metastring: meta

